json tags without commas leads to panic

Hi !
I tried ton generate the `avro` file for the following structure:

```
type User struct {
	Email string `avro:"email" json:"email"`
}
```

And unfortunetely, I had the following **panic**:

```
panic: runtime error: index out of range [1] with length 1 [recovered]
	panic: runtime error: index out of range [1] with length 1

goroutine 20 [running]:
testing.tRunner.func1(0xc000122400)
	c:/go/src/testing/testing.go:874 +0x3aa
panic(0x85df60, 0xc000075cc0)
	c:/go/src/runtime/panic.go:679 +0x1c0
github.com/heetch/avro.jsonFieldName(0x7d4967, 0x5, 0x0, 0x0, 0x927a60, 0x7f2f80, 0x7d496e, 0x19, 0x0, 0xc00012be38, ...)
	D:/dev/go/pkg/mod/github.com/heetch/avro@v0.0.0-20200115135924-a20461884c76/gotype.go:376 +0x1e7
github.com/heetch/avro.(*goTypeSchema).schemaForGoType(0xc000097a80, 0x927a60, 0x825300, 0x0, 0x0, 0x0, 0x0)
...
```

When I remove the second tag ("json" one), it succeed. 
So I think the issue is that when you have mutliple tags the lib does not handle properly tags array.

It should be an easy fix ? @rogpeppe 👍

--- a/analyze.go
+++ b/analyze.go
@@ -194,8 +194,9 @@ func (a *analyzer) eval(stack []int, path []pathElem) (retErr error) {
 			var enter func(v reflect.Value) (reflect.Value, bool)
 			switch elem.ftype.Kind() {
 			case reflect.Struct:
+				fieldIndex := info.fieldIndex
 				enter = func(v reflect.Value) (reflect.Value, bool) {
-					return v.Field(index), true
+					return v.Field(fieldIndex), true
 				}
 			case reflect.Interface:
 				enter = func(v reflect.Value) (reflect.Value, bool) {

--- a/encode.go
+++ b/encode.go
@@ -151,14 +151,17 @@ func typeEncoderUncached(names *Names, at schema.AvroType, t reflect.Type, info 
 			if len(info.entries) != len(def.Fields()) {
 				return errorEncoder(fmt.Errorf("entry count mismatch (info entries %d vs definition fields %d; %s vs %s)", len(info.entries), len(def.Fields()), t, def.Name()))
 			}
-			if t.NumField() != len(def.Fields()) {
-				return errorEncoder(fmt.Errorf("field count mismatch (%d vs %d; %s vs %s)", t.NumField(), len(def.Fields()), t, def.Name()))
-			}
-			fields := make([]encoderFunc, len(def.Fields()))
+			fieldEncoders := make([]encoderFunc, len(def.Fields()))
+			indexes := make([]int, len(def.Fields()))
 			for i, f := range def.Fields() {
-				fields[i] = typeEncoder(names, f.Type(), t.Field(i).Type, info.entries[i])
+				fieldIndex := info.entries[i].fieldIndex
+				fieldEncoders[i] = typeEncoder(names, f.Type(), t.Field(fieldIndex).Type, info.entries[i])
+				indexes[i] = fieldIndex
 			}
-			return structEncoder{fields}.encode
+			return structEncoder{
+				fieldEncoders: fieldEncoders,
+				fieldIndexes:  indexes,
+			}.encode
 		case *schema.EnumDefinition:
 			return longEncoder
 		case *schema.FixedDefinition:
@@ -373,12 +376,13 @@ func stringEncoder(e *encodeState, v reflect.Value) {
 }
 
 type structEncoder struct {
-	fields []encoderFunc
+	fieldIndexes  []int
+	fieldEncoders []encoderFunc
 }
 
 func (se structEncoder) encode(e *encodeState, v reflect.Value) {
-	for i, enc := range se.fields {
-		enc(e, v.Field(i))
+	for i, index := range se.fieldIndexes {
+		se.fieldEncoders[i](e, v.Field(index))
 	}
 }
 

--- a/gotype.go
+++ b/gotype.go
@@ -215,6 +215,9 @@ func (gts *goTypeSchema) schemaForGoType(t reflect.Type) (interface{}, error) {
 			// TODO  experiment by making optional only the fields that
 			// specify omitempty.
 			name, _ := jsonFieldName(f)
+			if name == "" {
+				continue
+			}
 			ftype, err := gts.schemaForGoType(f.Type)
 			if err != nil {
 				return nil, err
@@ -401,7 +404,7 @@ func jsonFieldName(f reflect.StructField) (name string, omitEmpty bool) {
 	switch {
 	case parts[0] == "":
 		return f.Name, omitEmpty
-	case parts[1] == "-":
+	case parts[0] == "-":
 		return "", omitEmpty
 	}
 	return parts[0], omitEmpty

--- a/gotype_test.go
+++ b/gotype_test.go
@@ -2,6 +2,7 @@ package avro_test
 
 import (
 	"encoding/json"
+	"sync"
 	"testing"
 	"time"
 
@@ -18,22 +19,98 @@ func TestSimpleGoType(t *testing.T) {
 			A: 1,
 			B: 2,
 		})
-		c.Assert(err, qt.Equals, nil)
+		if !c.Check(err, qt.Equals, nil) {
+			return
+		}
 		type TestRecord struct {
 			B int
 			A int
 		}
 		var x TestRecord
 		_, err = avro.Unmarshal(data, &x, wType)
-		c.Assert(err, qt.Equals, nil)
-		c.Assert(x, qt.Equals, TestRecord{
+		if !c.Check(err, qt.Equals, nil) {
+			return
+		}
+		c.Check(x, qt.Equals, TestRecord{
 			A: 1,
 			B: 2,
 		})
 	}
-	// Run the test twice to test caching.
-	test(t)
-	test(t)
+	// Run the test twice concurrently to test caching and potential race conditions.
+	var wg sync.WaitGroup
+	for i := 0; i < 3; i++ {
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			test(t)
+		}()
+	}
+	wg.Wait()
+}
+
+func TestGoTypeWithOmittedFields(t *testing.T) {
+	c := qt.New(t)
+	type R struct {
+		omit1 int
+		A     int
+		omit2 int
+		Omit3 int `json:"-"`
+		B     string
+	}
+	data, wType, err := avro.Marshal(R{
+		A: 1,
+		B: "hello",
+	})
+	c.Assert(err, qt.Equals, nil)
+	c.Assert(wType.String(), qt.JSONEquals, json.RawMessage(`{
+		"type": "record",
+		"name": "R",
+		"fields": [{
+			"default": 0,
+			"name": "A",
+			"type": "long"
+		}, {
+			"default": "",
+			"name": "B",
+			"type": "string"
+		}]
+	}`))
+
+	var r R
+	_, err = avro.Unmarshal(data, &r, wType)
+	c.Assert(err, qt.Equals, nil)
+	c.Assert(r, qt.Equals, R{A: 1, B: "hello"})
+}
+
+func TestGoTypeWithJSONTags(t *testing.T) {
+	c := qt.New(t)
+	type R struct {
+		A int    `json:"something"`
+		B string `json:"other,omitempty"`
+	}
+	data, wType, err := avro.Marshal(R{
+		A: 1,
+		B: "hello",
+	})
+	c.Assert(err, qt.Equals, nil)
+	c.Assert(wType.String(), qt.JSONEquals, json.RawMessage(`{
+		"type": "record",
+		"name": "R",
+		"fields": [{
+			"default": 0,
+			"name": "something",
+			"type": "long"
+		}, {
+			"default": "",
+			"name": "other",
+			"type": "string"
+		}]
+	}`))
+
+	var r R
+	_, err = avro.Unmarshal(data, &r, wType)
+	c.Assert(err, qt.Equals, nil)
+	c.Assert(r, qt.Equals, R{A: 1, B: "hello"})
 }
 
 func TestGoTypeWithTime(t *testing.T) {

--- a/typeinfo.go
+++ b/typeinfo.go
@@ -15,6 +15,10 @@ type azTypeInfo struct {
 	// ftype holds the Go type used for this Avro type (or nil for null).
 	ftype reflect.Type
 
+	// fieldIndex holds the index of the field if this entry is about
+	// a struct field.
+	fieldIndex int
+
 	// makeDefault is a function that returns the default
 	// value for a field, or nil if there is no default value.
 	makeDefault func() interface{}
@@ -37,7 +41,7 @@ func newAzTypeInfo(t reflect.Type) (azTypeInfo, error) {
 	case reflect.Struct:
 		info := azTypeInfo{
 			ftype:   t,
-			entries: make([]azTypeInfo, t.NumField()),
+			entries: make([]azTypeInfo, 0, t.NumField()),
 		}
 		// Note that RecordInfo is defined in such a way that
 		// the zero value gives useful defaults for a normal Go
@@ -57,6 +61,9 @@ func newAzTypeInfo(t reflect.Type) (azTypeInfo, error) {
 				// TODO consider struct embedding.
 				return azTypeInfo{}, fmt.Errorf("anonymous fields not supported")
 			}
+			if name, _ := jsonFieldName(f); name == "" {
+				continue
+			}
 			var required bool
 			var makeDefault func() interface{}
 			var unionVals []interface{}
@@ -69,11 +76,11 @@ func newAzTypeInfo(t reflect.Type) (azTypeInfo, error) {
 			if i < len(r.Unions) {
 				unionVals = r.Unions[i]
 			}
-			entry, err := newAzTypeInfoFromField(f.Type, required, makeDefault, unionVals)
+			entry, err := newAzTypeInfoFromField(f, required, makeDefault, unionVals)
 			if err != nil {
 				return azTypeInfo{}, err
 			}
-			info.entries[i] = entry
+			info.entries = append(info.entries, entry)
 		}
 		debugf("-> record, %d entries", len(info.entries))
 		return info, nil
@@ -86,7 +93,8 @@ func newAzTypeInfo(t reflect.Type) (azTypeInfo, error) {
 	}
 }
 
-func newAzTypeInfoFromField(t reflect.Type, required bool, makeDefault func() interface{}, unionVals []interface{}) (azTypeInfo, error) {
+func newAzTypeInfoFromField(f reflect.StructField, required bool, makeDefault func() interface{}, unionVals []interface{}) (azTypeInfo, error) {
+	t := f.Type
 	if t.Kind() == reflect.Ptr && len(unionVals) == 0 {
 		// It's a pointer but there's no explicit union entry, which means that
 		// the union defaults to ["null", type]
@@ -117,6 +125,7 @@ func newAzTypeInfoFromField(t reflect.Type, required bool, makeDefault func() in
 	}
 	info := azTypeInfo{
 		ftype:       t,
+		fieldIndex:  f.Index[0],
 		makeDefault: makeDefault,
 	}
 	if len(unionVals) == 0 {

